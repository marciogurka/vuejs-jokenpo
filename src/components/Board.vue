<template>
  <div class="column board-column">
    <div class="box">
      <div class="results">
        <div>
          <font-awesome-icon :icon="playerOption.icon"></font-awesome-icon>
          <span>{{ playerOption.name }}</span>
        </div>
        <h5>VS</h5>
        <div>
          <font-awesome-icon :icon="computerOption.icon"></font-awesome-icon>
          <span>{{ computerOption.name }}</span>
        </div>
      </div>
      <div class="info-message tile notification" v-bind:class="classObject">
        {{ checkMessage() }}
      </div>
    </div>
  </div>
</template>

<script>
import Menu from '@/components/Menu';

export const GameTexts = {
  waitingPlayer: 'Waiting for the player to choose',
  playerWin: 'You won!',
  playerDefeat: 'You lose!',
  waiting: 'Waiting...',
  tie: 'It\'s a tie!',
};

export default {
  name: 'Board',
  components: { Menu },
  props: ['playerOption', 'computerOption', 'winner'],
  data() {
    return {
      message: GameTexts.waitingPlayer,
    };
  },
  computed: {
    classObject() {
      return {
        'is-success': this.winner.indexOf('p1') > -1,
        'is-danger': this.winner.indexOf('p2') > -1,
        'is-warning': this.winner.indexOf('tie') > -1,
      };
    },
  },
  methods: {
    checkMessage() {
      if (this.winner) {
        switch (this.winner) {
          case 'p1':
            this.message = GameTexts.playerWin;
            break;
          case 'p2':
            this.message = GameTexts.playerDefeat;
            break;
          default:
            this.message = GameTexts.tie;
        }
      } else {
        this.message = GameTexts.waitingPlayer;
      }
      return this.message;
    },
  },
};

</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped>
.board-column {
  height: auto;
  display: flex;
  padding: 0.75rem;
}

@media screen and (max-width: 768px) {
  .board-column {
    padding: 0.75rem 1.5rem;
  }
}

.board-column > .box {
  width: 100%;
  flex-direction: column;
  align-content: space-around;
  justify-content: space-between;
  display: flex;
}

.results {
    display: flex;
    flex-direction: row;
    justify-content: center;
    align-items: center;
    height: 100%;
}

.results > div {
  display: flex;
  justify-content: center;
  flex-direction: column;
  align-items: center;
  padding: 15px;
  height: 100%;
}

.results > div > svg {
  width: 50%;
  height: auto;
}

@media screen and (max-width: 425px) {
  .results > div > svg {
    width: 90%;
    height: auto;
  }
}

.result > div > h5 {
  font-size: larger;
}

.info-message {
  max-height: 50px;
  width: 100%;
  display: flex;
  justify-content: center;
}
</style>
