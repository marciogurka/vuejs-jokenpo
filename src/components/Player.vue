<template>
<div class="column is-one-quarter-desktop is-one-third-tablet player-column">
  <div class="box">
      <div class="title title-name">
        {{ (!gameHasStarted && !isComputer) ? 'Welcome!' : name }}
      </div>
      <div v-if="isComputer">
        <img src="../assets/bot.jpg" alt="Computer profile image"/>
      </div>
      <div class="field" v-if="!isComputer && !gameHasStarted">
        <div class="control">
          <label for="name"> Please write your name below </label>
          <input class="input is-info name-input" id="name" type="text"
          placeholder="Put your name here" v-model="name">
        </div>
        <div class="control">
          <a class="button is-medium is-fullwidth is-success submit-name"
            @click.prevent="startGame"
            :disabled="!name">
            Play a match
          </a>
        </div>
      </div>
      <div class="field has-addons options-player" v-if="!isComputer && gameHasStarted">
        <div class="play-title-div">
          <h6 class="title is-6 play-options-title">Choose your play:</h6>
        </div>
        <div class="play-options">
          <a class="button is-large lizard-opt" @click="playerChoose(playerOptions.lizard)">
            <span class="icon is-medium">
              <font-awesome-icon icon="hand-lizard"></font-awesome-icon>
            </span>
          </a>
          <a class="button is-large rock-opt" @click="playerChoose(playerOptions.rock)">
            <span class="icon is-medium">
              <font-awesome-icon icon="hand-rock"></font-awesome-icon>
            </span>
          </a>
          <a class="button is-large scissors-opt" @click="playerChoose(playerOptions.scissors)">
            <span class="icon is-medium">
              <font-awesome-icon icon="hand-scissors"></font-awesome-icon>
            </span>
          </a>
          <a class="button is-large spock-opt" @click="playerChoose(playerOptions.spock)">
            <span class="icon is-medium">
              <font-awesome-icon icon="hand-spock"></font-awesome-icon>
            </span>
          </a>
          <a class="button is-large paper-opt" @click="playerChoose(playerOptions.paper)">
            <span class="icon is-medium">
              <font-awesome-icon icon="hand-paper"></font-awesome-icon>
            </span>
          </a>
        </div>
      </div>
      <router-link to="/" exact class="button is-small is-fullwidth back-button" v-if="!isComputer">
        <span>Back to menu</span>
      </router-link>
  </div>
</div>
</template>

<script>
export default {
  name: 'Player',
  props: ['isComputer', 'gameHasStarted', 'playerOptions'],
  data() {
    return {
      name: 'Bot',
    };
  },
  methods: {
    startGame() {
      if (this.name) {
        this.$emit('startGame', this.name);
      }
    },
    playerChoose(option) {
      this.$emit('playerChoose', option);
    },
  },
};
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped>
.player-column {
  display: flex;
  height: auto;
  padding: 0.75rem;
}

@media screen and (max-width: 768px) {
  .player-column {
    padding: 0.75rem 1.5rem;
  }
}

.player-column > .box {
  width: 100%;
}

.title-name {
  min-height: 36px;
  font-size: larger;
  margin-bottom: 0px;
}

.name-input {
  margin-bottom: 10px;
}

.options-player {
  flex-flow: wrap;
  justify-content: center;
}

.options-player > .play-title-div {
  display: flex;
  justify-content: center;
  align-items: center;
}

@media screen and (min-width: 425px) and (max-width: 768px) {
  .options-player h6 {
    margin: 0;
  }
}

.play-options-title {
  margin-bottom: 10px;
}

.play-options > a {
  margin: 5px;
}

.play-options {
  display: flex;
  flex-flow: wrap;
  justify-content: center;
}
</style>
