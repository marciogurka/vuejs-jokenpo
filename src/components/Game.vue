<template>
<div class="columns game">
  <Player :isComputer="false"
    :playerOptions="playerOptions"
    v-bind:gameHasStarted="gameHasStarted"
    @startGame="startGame"
    @playerChoose="playerChoose">
  </Player>
  <Board v-if="gameHasStarted"
    v-bind:playerOption="playerOption"
    v-bind:computerOption="computerOption"
    v-bind:winner="winner">
  </Board>
  <Player v-if="gameHasStarted" :isComputer="true"></Player>
</div>
</template>

<script>
import Player from '@/components/Player';
import Board from '@/components/Board';

const playerOptions = {
  lizard: {
    value: 0,
    icon: 'hand-lizard',
    name: 'Lizard',
  },
  rock: {
    value: 1,
    icon: 'hand-rock',
    name: 'Rock',
  },
  scissors: {
    value: 2,
    icon: 'hand-scissors',
    name: 'Scissors',
  },
  spock: {
    value: 3,
    icon: 'hand-spock',
    name: 'Spock',
  },
  paper: {
    value: 4,
    icon: 'hand-paper',
    name: 'Paper',
  },
};

export default {
  name: 'Game',
  components: {
    Player,
    Board,
  },
  data() {
    return {
      gameHasStarted: false,
      playerOption: {
        icon: 'question',
      },
      computerOption: {
        icon: 'question',
      },
      winner: '',
      playerOptions,
    };
  },
  methods: {
    startGame() {
      this.gameHasStarted = true;
      // reseting the board icons
      this.playerOption = {
        icon: 'question',
      };
      this.computerOption = {
        icon: 'question',
      };
      this.winner = '';
    },
    playerChoose(option) {
      this.playerOption = option;
      const botOption = this.generateComputerOption();
      this.computerOption = botOption;
      this.winner = this.checkWinner(option.value, botOption.value);
    },
    generateComputerOption() {
      const computerOptValue = Math.floor(Math.random() * 4);
      return Object.keys(playerOptions)
        .map(option => playerOptions[option])
        .find(option => option.value === computerOptValue);
    },
    checkWinner(playerOpt, computerOpt) {
      if (playerOpt === computerOpt) {
        return 'tie';
      } else if (playerOpt === undefined || playerOpt === null) {
        throw new Error('Opção inválida! Tente Novamente');
      } else if (computerOpt === undefined || computerOpt === null) {
        throw new Error('Opção inválida! Tente Novamente');
      }

      switch (playerOpt) {
        case 0:
          if (Math.abs(playerOpt - computerOpt) > 2) {
            return 'p1';
          }
          return 'p2';
        case 1:
          if (Math.abs(playerOpt - computerOpt) <= 1) {
            return 'p1';
          }
          return 'p2';
        case 2:
          if (Math.abs(playerOpt - computerOpt) === 2) {
            return 'p1';
          }
          return 'p2';
        case 3:
          switch (computerOpt) {
            case 1:
            case 2:
              return 'p1';
            case 0:
            case 4:
              return 'p2';
            default:
              throw new Error('Opção inválida! Tente Novamente');
          }
        case 4:
          if ((playerOpt - computerOpt) % 2 !== 0) {
            return 'p1';
          }
          return 'p2';
        default:
          throw new Error('Opção inválida! Tente Novamente');
      }
    },
  },
};
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped>
.game {
  justify-content: center;
}
</style>
