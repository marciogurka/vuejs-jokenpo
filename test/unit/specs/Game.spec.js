import { mount, RouterLinkStub } from '@vue/test-utils';
import Game from '@/components/Game';
import Player from '@/components/Player';
import Board from '@/components/Board';
import { library } from '@fortawesome/fontawesome-svg-core';
import { faHandLizard, faHandRock, faHandPaper, faHandSpock, faHandScissors, faQuestion } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome';

library.add(faHandLizard, faHandRock, faHandPaper, faHandSpock, faHandScissors, faQuestion);

const mockMath = Object.create(global.Math);
mockMath.random = () => 1;
global.Math = mockMath;

describe('GameSpec.vue', () => {
  let cmp;
  // stubs object to avoid error rendering custom elements
  const stubs = {
    RouterLink: RouterLinkStub,
    'font-awesome-icon': FontAwesomeIcon,
  };
  // Method to create component with custom props
  const createCmp = propsData => mount(Game, { propsData, stubs });
  /* Mouting the component before each test */
  beforeEach(() => {
    cmp = createCmp({});
  });

  it('has the expected html structure', () => {
    expect(cmp.element).toMatchSnapshot();
  });

  it('is a Vue Component', () => {
    expect(cmp.isVueInstance()).toBe(true);
  });

  it('is a Game component', () => {
    expect(cmp.is(Game)).toBe(true);
  });

  describe('Before the game starts', () => {
    /* Mouting the component before each test */
    beforeEach(() => {
      cmp = createCmp({});
    });
    it('has a Player component', () => {
      expect(cmp.contains(Player)).toBe(true);
    });

    it('has not a Board component', () => {
      expect(cmp.contains(Board)).toBe(false);
    });

    it('has only one Player component being displayed', () => {
      expect(cmp.findAll(Player).length).toBe(1);
    });

    it('has only one Player that is not a bot component being displayed', () => {
      expect(cmp.findAll(Player).at(0).props().isComputer).toBe(false);
    });

    it('Calls startGame when @startGame happens', () => {
      const stub = jest.fn();
      cmp.setMethods({ startGame: stub });

      cmp.find(Player).vm.$emit('startGame', 'Bot');
      expect(stub).toBeCalledWith('Bot');
    });
  });

  describe('When the game starts', () => {
    /* Mouting the component before each test */
    beforeEach(() => {
      cmp = createCmp({
        playerOption: {
          icon: 'question',
        },
        computerOption: {
          icon: 'question',
        },
      });
      cmp.vm.gameHasStarted = true;
    });

    it('this.winner is empty when you start a game the checkResult method', () => {
      expect(cmp.vm.$data.winner).toBe('');
    });

    it('playerOption has default value', () => {
      expect(cmp.vm.$data.playerOption).toEqual({ icon: 'question' });
    });

    it('computerOption has default value', () => {
      expect(cmp.vm.$data.computerOption).toEqual({ icon: 'question' });
    });

    it('has 2 Player components being displayed', () => {
      expect(cmp.findAll(Player).length).toBe(2);
    });

    it('has a Player component that is not a bot', () => {
      expect(cmp.findAll(Player).at(0).props().isComputer).toBe(false);
    });

    it('has a Player component that is a bot', () => {
      expect(cmp.findAll(Player).at(1).props().isComputer).toBe(true);
    });

    it('has a Board component', () => {
      expect(cmp.contains(Board)).toBe(true);
    });

    describe('And the user choose an option', () => {
      it('calls playerChoose when @playerChoose happens', () => {
        const stub = jest.fn();
        cmp.setMethods({ playerChoose: stub });
        cmp.findAll(Player).at(0).vm.$emit('playerChoose');
        expect(stub).toHaveBeenCalled();
      });

      describe('And its a lizard option', () => {
        it('calls playerChoose with lizard option when @playerChoose with lizard option happens', () => {
          const stub = jest.fn();
          cmp.setMethods({ playerChoose: stub });
          cmp.find(Player).find('.lizard-opt').trigger('click');
          expect(stub).toBeCalledWith(cmp.vm.$data.playerOptions.lizard);
        });

        it('and bot choose lizard, its a tie', () => {
          expect(cmp.vm.checkWinner(0, 0)).toBe('tie');
        });

        it('and bot choose rock, its a bot win', () => {
          expect(cmp.vm.checkWinner(0, 1)).toBe('p2');
        });

        it('and bot choose scissors, its a bot win', () => {
          expect(cmp.vm.checkWinner(0, 2)).toBe('p2');
        });

        it('and bot choose spock, its a player win', () => {
          expect(cmp.vm.checkWinner(0, 3)).toBe('p1');
        });

        it('and bot choose paper, its a player win', () => {
          expect(cmp.vm.checkWinner(0, 4)).toBe('p1');
        });
      });

      describe('And its a rock option', () => {
        it('calls playerChoose with rock option when @playerChoose with rock option happens', () => {
          const stub = jest.fn();
          cmp.setMethods({ playerChoose: stub });
          cmp.find(Player).find('.rock-opt').trigger('click');
          expect(stub).toBeCalledWith(cmp.vm.$data.playerOptions.rock);
        });

        it('and bot choose rock, its a tie', () => {
          expect(cmp.vm.checkWinner(1, 1)).toBe('tie');
        });

        it('and bot choose lizard, its a player win', () => {
          expect(cmp.vm.checkWinner(1, 0)).toBe('p1');
        });

        it('and bot choose scissors, its a player win', () => {
          expect(cmp.vm.checkWinner(1, 2)).toBe('p1');
        });

        it('and bot choose spock, its a bot win', () => {
          expect(cmp.vm.checkWinner(1, 3)).toBe('p2');
        });

        it('and bot choose paper, its a bot win', () => {
          expect(cmp.vm.checkWinner(1, 4)).toBe('p2');
        });
      });

      describe('And its a scissors option', () => {
        it('calls playerChoose with scissors option when @playerChoose with scissors option happens', () => {
          const stub = jest.fn();
          cmp.setMethods({ playerChoose: stub });
          cmp.find(Player).find('.scissors-opt').trigger('click');
          expect(stub).toBeCalledWith(cmp.vm.$data.playerOptions.scissors);
        });

        it('and bot choose scissors, its a tie', () => {
          expect(cmp.vm.checkWinner(2, 2)).toBe('tie');
        });

        it('and bot choose lizard, its a player win', () => {
          expect(cmp.vm.checkWinner(2, 0)).toBe('p1');
        });

        it('and bot choose rock, its a bot win', () => {
          expect(cmp.vm.checkWinner(2, 1)).toBe('p2');
        });

        it('and bot choose spock, its a bot win', () => {
          expect(cmp.vm.checkWinner(2, 3)).toBe('p2');
        });

        it('and bot choose paper, its a player win', () => {
          expect(cmp.vm.checkWinner(2, 4)).toBe('p1');
        });
      });
      describe('And its a spock option', () => {
        it('calls playerChoose with spock option when @playerChoose with spock option happens', () => {
          const stub = jest.fn();
          cmp.setMethods({ playerChoose: stub });
          cmp.find(Player).find('.spock-opt').trigger('click');
          expect(stub).toBeCalledWith(cmp.vm.$data.playerOptions.spock);
        });

        it('and bot choose spock, its a tie', () => {
          expect(cmp.vm.checkWinner(3, 3)).toBe('tie');
        });

        it('and bot choose lizard, its a bot win', () => {
          expect(cmp.vm.checkWinner(3, 0)).toBe('p2');
        });

        it('and bot choose paper, its a player win', () => {
          expect(cmp.vm.checkWinner(3, 1)).toBe('p1');
        });

        it('and bot choose scissors, its a player win', () => {
          expect(cmp.vm.checkWinner(3, 2)).toBe('p1');
        });

        it('and bot choose paper, its a bot win', () => {
          expect(cmp.vm.checkWinner(3, 4)).toBe('p2');
        });
      });
      describe('And its a paper option', () => {
        it('calls playerChoose with paper option when @playerChoose with paper option happens', () => {
          const stub = jest.fn();
          cmp.setMethods({ playerChoose: stub });
          cmp.find(Player).find('.paper-opt').trigger('click');
          expect(stub).toBeCalledWith(cmp.vm.$data.playerOptions.paper);
        });

        it('and bot choose paper, its a tie', () => {
          expect(cmp.vm.checkWinner(4, 4)).toBe('tie');
        });

        it('and bot choose lizard, its a bot win', () => {
          expect(cmp.vm.checkWinner(4, 0)).toBe('p2');
        });

        it('and bot choose rock, its a player win', () => {
          expect(cmp.vm.checkWinner(4, 1)).toBe('p1');
        });

        it('and bot choose scissors, its a bot win', () => {
          expect(cmp.vm.checkWinner(4, 2)).toBe('p2');
        });

        it('and bot choose spock, its a player win', () => {
          expect(cmp.vm.checkWinner(4, 3)).toBe('p1');
        });
      });
    });
  });
  describe('Functions tests', () => {
    beforeEach(() => {
      cmp = createCmp({
        playerOption: {
          icon: 'question',
        },
        computerOption: {
          icon: 'question',
        },
      });
      cmp.vm.gameHasStarted = true;
    });

    it('should throw an error if no player option is passed', () => {
      expect(() => { cmp.vm.checkWinner(null, 0); }).toThrow(new Error('Opção inválida! Tente Novamente'));
    });

    it('should throw an error if no bot option is passed', () => {
      expect(() => { cmp.vm.checkWinner(0, null); }).toThrow(new Error('Opção inválida! Tente Novamente'));
    });

    it('should throw an error if get a default option in player switch function', () => {
      expect(() => { cmp.vm.checkWinner(7, 1); }).toThrow(new Error('Opção inválida! Tente Novamente'));
    });

    it('should throw an error if get a default option in bot switch function', () => {
      expect(() => { cmp.vm.checkWinner(3, 6); }).toThrow(new Error('Opção inválida! Tente Novamente'));
    });

    it('generate a random number for the bot', () => {

    });

    it('this.winner is not null after the checkResult method', () => {

    });
  });
});
