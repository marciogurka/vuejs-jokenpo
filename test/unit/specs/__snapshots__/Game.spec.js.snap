// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameSpec.vue has the expected html structure 1`] = `
<div
  class="columns game"
>
  <div
    class="column is-one-quarter-desktop is-one-third-tablet player-column"
  >
    <div
      class="box"
    >
      <div
        class="title title-name"
      >
        
        Welcome!
      
      </div>
       
      <!---->
       
      <div
        class="field"
      >
        <div
          class="control"
        >
          <label
            for="name"
          >
             Please write your name below 
          </label>
           
          <input
            class="input is-info name-input"
            id="name"
            placeholder="Put your name here"
            type="text"
          />
        </div>
         
        <div
          class="control"
        >
          <a
            class="button is-medium is-fullwidth is-success submit-name"
          >
            
            Play a match
          
          </a>
        </div>
      </div>
       
      <!---->
       
      <a
        class="button is-small is-fullwidth back-button"
      >
        <span>
          Back to menu
        </span>
      </a>
    </div>
  </div>
   
  <!---->
   
  <!---->
</div>
`;
